### PR TITLE
Visual Studio Code development environment.

### PR DESCRIPTION
**Pull request by hacktobeer**
This PR provides an easy Turbinia Visual Studio Code development environment and documentation.

* vscode container development environment (including all dependencies pre-installed for fast bootstrap)
* Launch profiles for server, workers and client
* pylint, yapf settings configured

### CODE REVIEW

File: docs/developer/develop-vscode.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+Start server
+```
+$ python3 turbinia/turbiniactl.py -S server
+```
+Start worker
+```
+$ python3 turbinia/turbiniactl.py -S celeryworker
+```
+Fetch and process some evidence
```
Comment (aarontp) [inline docs/developer/develop-vscode.md:52]: Maybe mention that the request ID is the in the output of the first turbiniactl.py command?

File: docs/developer/develop-vscode.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+You're done :). Go to Step 8 for a vscode debugging example.
+
+#### Step 6 - GCP hybrid Turbinia setup
+Follow the ‘GCP Setup’ section [here](https://turbinia.readthedocs.io/en/latest/user/install-manual.html) and setup Cloud Functions, a GCE bucket, Datastore and PubSub.
+* Create a pubsub topic ,eg ‘turbinia-dev’.
+* Create a GCE storage bucket with a unique name.
+* Create the GCP Cloud Functions
+```
+$ git clone https://github.com/forseti-security/osdfir-infrastructure.git
```
Comment (aarontp) [inline docs/developer/develop-vscode.md:70]: If this is only being used for the cloud function files, we should be able to take them out of the `turbinia/tools/gcf_init` in the Turbinia project/code.   Also, if you do the config setup first, you can replace all of this with a call to `turbinia/tools/gcf_init/deploy_gcf.py` which reads the config and does these same gcloud deployment commands.

File: docs/developer/develop-vscode.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+Edit the configuration file ```~/.turbiniarc``` and set below variables according to the GCP project you are using. Make sure all values are between quotes!
+```
+TURBINIA_PROJECT = '[your_gcp_project_name]' 
+TURBINIA_REGION = '[your_preferred_region]'  eg 'us-central1'
+TURBINIA_ZONE = '[your_preferred_zone]'  eg 'us-central1-f'
+PUBSUB_TOPIC = '[your_gcp_pubsub_topic_name]' eg 'turbinia-dev'
+BUCKET_NAME = '[your_gcp_bucket_name]'
+```
+
+Setup authentication for the GCP project.
```
Comment (aarontp) [inline docs/developer/develop-vscode.md:92]: I think you might want to move this to before the cloud function deployment.

File: docs/developer/develop-vscode.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+$ python3 turbinia/turbiniactl.py status -r [request_id]
+```
+
+This should process the evidence and show output in each terminal for server and worker. Results will be stored in ```/evidence```.
+
+You're done :). Go to Step 8 for a vscode debugging example.
+
+#### Step 6 - GCP hybrid Turbinia setup
+Follow the ‘GCP Setup’ section [here](https://turbinia.readthedocs.io/en/latest/user/install-manual.html) and setup Cloud Functions, a GCE bucket, Datastore and PubSub.
+* Create a pubsub topic ,eg ‘turbinia-dev’.
```
Comment (aarontp) [inline docs/developer/develop-vscode.md:66]: Couple minor nits in this section:  The comma and spacing on the line above, and also either end all of the bullet points with periods or remove them for consistency.

File: docs/developer/develop-vscode.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+```
+$ git clone https://github.com/forseti-security/osdfir-infrastructure.git
+$ cd osdfir-infrastructure
+$ export TURBINIA_REGION=us-central1
+$ gcloud -q functions deploy gettasks --region $TURBINIA_REGION --source modules/turbinia/data/ --runtime nodejs10 --trigger-http --memory 256MB --timeout 60s
+$ gcloud -q functions deploy closetask --region $TURBINIA_REGION --source modules/turbinia/data/ --runtime nodejs10 --trigger-http --memory 256MB --timeout 60s
+$ gcloud -q functions deploy closetasks  --region $TURBINIA_REGION --source modules/turbinia/data/ --runtime nodejs10 --trigger-http --memory 256MB --timeout 60s
+```
+
+Create the Turbinia hybrid configuration file.
```
Comment (aarontp) [inline docs/developer/develop-vscode.md:78]: This should be done in the vs code terminals as well, right?

File: docs/developer/develop-vscode.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+BUCKET_NAME = '[your_gcp_bucket_name]'
+```
+
+Setup authentication for the GCP project.
+```
+$ gcloud auth login
+$ gcloud auth application-default login
+```
+
+#### Step 7 - GCP hybrid Turbinia verification
```
Comment (aarontp) [inline docs/developer/develop-vscode.md:98]: I think step 5 and 7 are the same (maybe with the exception of the GCS bucket comment).  Maybe we combine them?

File: docs/developer/develop-vscode.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+```
+Terminal 3 - Fetch and process some evidence
+```
+$ curl https://raw.githubusercontent.com/obsidianforensics/hindsight/master/tests/fixtures/profiles/60/History > History
+$ tar -vzcf /evidence/history.tgz History
+$ python3 turbinia/turbiniactl.py compresseddirectory -l /evidence/history.tgz 
+$ python3 turbinia/turbiniactl.py -a status -r [request_id]
+```
+
+This should process the evidence and show output in each terminal for server and worker. Results will be stored in ```/evidence``` and in the GCE bucket.
```
Comment (aarontp) [inline docs/developer/develop-vscode.md:117]: s/GCE bucket/GCS bucket/

File: docs/developer/develop-vscode.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+You're done :). Go to Step 8 for a vscode debugging example.
+
+#### Step 8 - Debugging example
+When you are developing code you want to be able to step through your code and inspect variables. We can do this by running the different launch profiles provided. Visual Studio Code launch profiles are provided for server, celeryworker, psqworker and client requests. 
+
+As a small example we will change the version string of Turbinia.
+* edit ```turbinia/__init__.py``` and change around line 24 the version string from "unknown" to "iamrad" (The person who wrote this is old...).
```
Comment (aarontp) [inline docs/developer/develop-vscode.md:125]: Minor nits:  Maybe start the bullet point lines with capitals for consistency.

**Comment by hacktobeer** [comment]
PTAL

File: docs/developer/develop-vscode.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+#### Step 3 - Open in Visual Studio Code
+Open the folder in vscode and choose to "Reopen in Container" when asked (vscode will see the ```.devcontainer``` folder in the turbinia cloned source tree). Vscode will build your Turbinia development container and that will take a couple of minutes.
+
+Note: if vscode does not ask you to reopen in a container you need to verify you have installed the Remote Development extension!
+
+When this is finished continue with Step 4 for a local Turbinia setup or Step 6 for a GCP hybrid setup. 
+
+#### Step 4 - Local Turbinia setup
+The local turbinia setup will use redis and celery. Let's create the configuration file for this setup.
```
[the remark this answers is not in the document]
Comment (hacktobeer) [reply docs/developer/develop-vscode.md:46]: Good one, added clarifications to all shell command instructions (this is something people will overlook fast)

**Review by aarontp — APPROVED** [review]
LGTM

**Comment by aarontp** [comment]
LGTM, but for some reason it thinks there are conflicts in `.github/workflows/e2e.yml`, and it won't let me resolve them through the web UI.

**Comment by hacktobeer** [comment]
That.s weird as that file is not even in this PR....let me check.

**Comment by hacktobeer** [comment]
Did some git magic and resolved it.

**Pull request merged**